samza-core/src/main/java/org/apache/samza/storage/blobstore/BlobStoreRestoreManager.java

```diff
@@ -119,13 +119,18 @@ public BlobStoreRestoreManager(TaskModel taskModel, ExecutorService restoreExecu
 
   @Override
   public void init(Checkpoint checkpoint) {
+    init(checkpoint, false);
+  }
+
+  public void init(Checkpoint checkpoint, Boolean getDeletedBlob) {
     long startTime = System.nanoTime();
     LOG.debug("Initializing blob store restore manager for task: {}", taskName);
 
     blobStoreManager.init();
 
     // get previous SCMs from checkpoint
-    prevStoreSnapshotIndexes = blobStoreUtil.getStoreSnapshotIndexes(jobName, jobId, taskName, checkpoint, storesToRestore);
+    prevStoreSnapshotIndexes =
+        blobStoreUtil.getStoreSnapshotIndexes(jobName, jobId, taskName, checkpoint, storesToRestore, getDeletedBlob);
```

Comment:
Why not always get snapshot index with getDeletedBlob = true in init (and remove this variant)?

IIUC (correct me if wrong):
1. We only getStoreSnapshotIndex twice during container startup, once in BSRestoreManager, once in BSBackupManager.
2. Either both requests with getDeleted will succeed, or not.
        a. If both succeed, then regardless of whether the snapshot index blob is deleted or not, restore and future backups will continue uninterrupted. If the restore for an actual file in the snapshot fails, then BSRM will retry the entire restore() with getDeletedBlob = true with the same snapshot index.
        b. If either fails, then the snapshot index blob is gone for good and there is nothing we can do.
        
 So why do we need the let the first BSRM init request fail for deleted snapshot index blob?

Reply:
Discussed offline. We use "snapshot index is deleted" as a indicator for "files are deleted" (since that's the last step in the post-commit cleanup). Making this check upfront allows us to keep the shouldRestore optimization below for the happy path. If snapshot index is deleted, then all files need to be restored with getDeletedBlob = true and restored, regardless of whether there is a local checkpoint or not. Resolving this.